### PR TITLE
How to deploy to App Service

### PR DESCRIPTION
**Pull request by rayluo**
Hi, @pamelafox , this web app sample itself was already capable to be deployed to App Service. But there are quite some tricky traps during the configuration. I added some know-how based on my recent attempt. Please help review and [proofread](https://github.com/Azure-Samples/ms-identity-python-webapp/tree/app-service-howto#readme).

Let me know if you would also want to update some corresponding docs that you recently worked on. I can then add a new tag for you.

### CODE REVIEW

File: .env.sample.b2c [hunk 1]
```diff
@@ -1,3 +1,6 @@
+# Note: If you are using Azure App Service, go to your app's Configuration,
+# and then set the following values into your app's "Application settings".
```
Comment (pamelafox) [inline .env.sample.b2c:2]: This message is misleading since FLASK_DEBUG is also in here, and they most definitely shouldn't set that to true. Can you put the comment above the registration secrets to make it clear it's not about FLASK_DEBUG? (Generally, it's common for env variables locally to be different in production, like for database connection strings, so they may add more env variables that would be local-only in future)

Alternatively, that comment can stay in the "Deployment" tutorial

File: README.md [hunk 2]
```diff
@@ -17,6 +17,28 @@ To get started with this sample, you have two options:
...
+Here we will describe how to deploy it to
+[Azure App Service](https://azure.microsoft.com/en-us/products/app-service).
+
+* Follow the ["Quickstart: Deploy a Python (Django or Flask) web app to Azure App Service"](https://learn.microsoft.com/en-us/azure/app-service/quickstart-python),
+  but replace its sample app (which does not do user sign-in) with this web app.
+
+* In particular, if you choose to ["deploy using Local Git" in "step 3 - Deploy your application code to Azure"](https://learn.microsoft.com/en-us/azure/app-service/quickstart-python?tabs=flask%2Cwindows%2Cazure-cli%2Clocal-git-deploy%2Cdeploy-instructions-azportal%2Cterminal-bash%2Cdeploy-instructions-zip-azcli#3---deploy-your-application-code-to-azure),
+  an [application-scope credential](https://learn.microsoft.com/en-us/azure/app-service/deploy-configure-credentials?tabs=portal#appscope)
+  will be automatically created with the shape as `your_app_name\$your_app_name`.
+  But your actual git username is only the `$your_app_name` part.
```
Comment (pamelafox) [inline README.md:33]: Is that a bug with the tutorial? Or somehow unique to this application? (If a bug with the tutorial, we should fix it there)

File: README.md [hunk 3]
```diff
@@ -17,6 +17,28 @@ To get started with this sample, you have two options:
...
+
+* Do not forget to setup
+  [several environment variables](https://github.com/Azure-Samples/ms-identity-python-webapp/blob/master/.env.sample)
+  needed by this sample.
+  Otherwise your website will usually render a blank "Internal Error" page
+  (Flask's debug mode output does not seem to be relayed by App Service).
+
+
```
Comment (pamelafox) [inline README.md:41]: The debug mode output is not relayed as it's bad practice to set FLASK_DEBUG to True on production. In the case of "Internal Error", developers can check the Log Stream for details.

File: README.md [hunk 4]
```diff
@@ -17,6 +17,28 @@ To get started with this sample, you have two options:
...
+* In particular, if you choose to ["deploy using Local Git" in "step 3 - Deploy your application code to Azure"](https://learn.microsoft.com/en-us/azure/app-service/quickstart-python?tabs=flask%2Cwindows%2Cazure-cli%2Clocal-git-deploy%2Cdeploy-instructions-azportal%2Cterminal-bash%2Cdeploy-instructions-zip-azcli#3---deploy-your-application-code-to-azure),
+  an [application-scope credential](https://learn.microsoft.com/en-us/azure/app-service/deploy-configure-credentials?tabs=portal#appscope)
+  will be automatically created with the shape as `your_app_name\$your_app_name`.
+  But your actual git username is only the `$your_app_name` part.
+
+* Do not forget to setup
+  [several environment variables](https://github.com/Azure-Samples/ms-identity-python-webapp/blob/master/.env.sample)
+  needed by this sample.
+  Otherwise your website will usually render a blank "Internal Error" page
+  (Flask's debug mode output does not seem to be relayed by App Service).
```
Comment (pamelafox) [inline README.md:39]: I suggest making it clear how to set them up, like:

In the App Service page, in the left menu, select **Configuration**.

In the **Application settings** tab, select **New application setting**. Name the setting `CLIENT_ID`. Paste the value from the *.env* file. Select OK.

Repeat for `CLIENT_SECRET` and `TENANT_ID` (if using it).

File: app.py [hunk 5]
```diff
@@ -48,6 +48,13 @@ def logout():
 
 @app.route("/")
 def index():
+    if not (app.config["CLIENT_ID"] and app.config["CLIENT_SECRET"]):
```
Comment (pamelafox) [inline app.py:51]: Could this return templates/missing_config.hml and put the HTML in there? It's worrisome when I see HTML inside a Python file, and it's also invalid HTML since it's lacking doctype, html, head, title, body, etc?

Also, I see you're saying `not app.config["CLIENT_ID"]` versus `"CLIENT_ID" not in app.config`. Is that because it's set to an empty string at this point? If it's set to `None`, we'd need the latter check, I think.

**Review by pamelafox — CHANGES_REQUESTED** [review]
Nice to give them ideas for how to deploy! I'm looking into azd-ifying but it looks like there's a blocking issue at the moment (https://github.com/Azure/azure-dev/issues/1714) so this looks like a good approach.

File: .env.sample.b2c (hunk 1 above)
Comment (rayluo) [reply .env.sample.b2c:2]: I understand your point. For the sake of discussion, why was FLASK_DEBUG=True acceptable before this PR? We certainly anticipated app developers to subsequently deploy this sample to their hosting system and eventually make proper changes in the final production system.

Currently this paragraph is put at the very top of this file for visibility. I was bitten by the forgetting to setup App Service's application settings, and the troubleshooting experience was like shooting in the dark (those stream log does not seem to help - more on this later). I foresee the majority of future customers who attempts this sample on App Service would run into the same issue. We want to help them. "Burying" this paragraph in the middle of a file won't catch their attention.

How about, we do it the other way around? Let's comment out the FLASK_DEBUG=True line by default. And optionally you can modify the quickstart docs to append the `--debug` command line parameter in the "flask run ...". If you would agree on this direction, I can send out a different PR to bump our Flask dependency version to 2.2+ (the `--debug` is only available since Flask 2.2+).

File: README.md (hunk 2 above)
Comment (rayluo) [reply README.md:33]: It is not a bug in tutorial. In fact, the tutorial I quoted already gives a hint albeit buried deep inside the instruction.
But it has been an FAQ [here](https://github.com/MicrosoftDocs/azure-docs/issues/6548) and [there](https://learn.microsoft.com/en-us/answers/questions/870466/local-git-authentication-fail?page=1&orderby=helpful). And it can be said as a bug in the Portal UI. I reported it, and CCed you in the mail thread. Unless it is fixed in Portal, I suggest we leave this line here. It will save each future person a hour or two of frustrating fiddling.

File: README.md (hunk 3 above)
Comment (rayluo) [reply README.md:41]: Yeah, I figure they strip out all the error page content for security concern, and rightfully so.

Log Stream seems to contain deployment logs only, but no app level logs. Presumably because [Log Stream only captures output on stdout](https://learn.microsoft.com/en-us/azure/app-service/quickstart-python?tabs=flask%2Cwindows%2Cazure-cli%2Cvscode-deploy%2Cdeploy-instructions-azportal%2Cterminal-bash%2Cdeploy-instructions-zip-azcli#5---stream-logs). Who would write print() in a server side app? Our app does not.

Comment (pamelafox) [reply README.md:41]: I believe we could configure it to get logger.info() to output to Log Stream - but the default for Flask is likely warning level (vs info level). Instructions here:
https://flask.palletsprojects.com/en/2.2.x/logging/
Have you tried logging.warn()? Or app.logger.warn()?

File: app.py (hunk 5 above)
Comment (rayluo) [reply app.py:51]: > Could this return templates/missing_config.hml and put the HTML in there? It's worrisome when I see HTML inside a Python file, and it's also invalid HTML since it's lacking doctype, html, head, title, body, etc?

Due to the lack of the familiar Flask Debug output, and the lack of useful Log Stream (mentioned above), I find this little check-and-hint here tremendously helpful for the *very first run*. After that, the app developer would probably just remove these two lines. For that reason, I do not want to overdo it otherwise the develop would need to - but likely forget to - delete one more file. You know, [practicality beats purity](https://zen-of-python.info/although-practicality-beats-purity.html#9).

> Also, I see you're saying not app.config["CLIENT_ID"] versus "CLIENT_ID" not in app.config. Is that because it's set to an empty string at this point? If it's set to None, we'd need the latter check, I think.

It is not about the empty string. Our sample would set it to None by default. Yet it doesn't harm to be robust to guard against a potential empty string input.

But, regardless of the None-vs-empty-string situation, `"CLIENT_ID" not in app.config` is never the semantics we want here. We do not just want an existence of a key, we need some real value to be consume.

File: README.md (hunk 3 above)
Comment (rayluo) [reply README.md:41]: Fair point. I tried app.logger.error() just now. It technically works, and my log made it into the Log Stream. But it is not really helpful. It turns out, the Log Stream is an ever-increasing text page that combining *many* log files in the log folder. My error log line is just one line among other 500 lines in the stream, and my log line is not even guaranteed to be the last line being shown.

I still think that inline, to-the-point, error message inside our index() is much more helpful. Its discoverability is superb. First-time app service users may not even find the page for Log Stream, but they won't miss out an error served on their web app's index page.

File: README.md (hunk 4 above)
Comment (rayluo) [reply README.md:39]: Since we worked out the new README file (thanks again, Pamel!), I am convinced that we want a slim README here. So, I would like to avoid those "open this page, select that tab, click the button, tap OK" style of instruction.

Admittedly, the paragraph here "do not forget ... otherwise ..." might not sound professional. How about I change it into something like this?

> [Configure your app's setting](https://learn.microsoft.com/en-us/azure/app-service/configure-common?tabs=portal#configure-app-settings) to define [these environment variables](https://github.com/Azure-Samples/ms-identity-python-webapp/blob/master/.env.sample).

File: .env.sample.b2c (hunk 1 above)
Comment (pamelafox) [reply .env.sample.b2c:2]: Good idea! I didn't realize the FLASK_DEBUG risk as all of the other samples I've worked on clearly use .env for local-only variables, and set the env variables to different values for prod (like for database connections). This ones different since you want to connect locally to the actual API.

I sent PR for debug command: https://github.com/MicrosoftDocs/azure-docs-pr/pull/232094

File: README.md (hunk 4 above)
Comment (pamelafox) [reply README.md:39]: That more concise wording is fine, as long as we keep .env only to production-needed variables. I made that change in my PR-to-this-PR: https://github.com/Azure-Samples/ms-identity-python-webapp/pull/98

File: README.md (hunk 3 above)
Comment (pamelafox) [reply README.md:41]: Lol, Log Stream is pretty janky right now, I tell the App Service team that every week- they say an improvement is coming! When I need legit logs that aren't strangely ordered, I go to Advanced Tools and download the Docker logs.

Anyway, yep, I like your idea, I just want it to be a template. Made change in my PR: https://github.com/Azure-Samples/ms-identity-python-webapp/pull/98

File: README.md (hunk 2 above)
Comment (pamelafox) [reply README.md:33]: Okay, I need to give it a go myself to wrap my head around the bug as I don't remember it, thanks for raising to the team. Leaving the extra note here is fine.

File: app.py (hunk 5 above)
Comment (pamelafox) [reply app.py:51]: I was worried when I saw dict access since typically a not-found-key raises a KeyError, but I realize now that app_config will set CLIENT_ID to None if its not found, so CLIENT_ID will always be a key. So this logic is correct, sorry for false alarm.

File: README.md (hunk 2 above)
Comment (pamelafox) [reply README.md:33]: Just did the flow, I see what you mean. There's another issue with the flow in that the Quickstart presumes the branch is called "main" locally but this repo has it called "master". Can you rename it? I did it for my old repo last week and GitHub makes it pretty easy to rename in the settings. Most samples use "main" that I've seen.

File: app.py (hunk 5 above)
Comment (rayluo) [reply app.py:51]: > I was worried when I saw dict access since typically a not-found-key raises a KeyError

Understood. To guard against that KeyError yet _still_ check the value (rather than just checking the key), the proper way is to `if dictionary.get("KEY")` rather than `if "KEY" in dictionary`.

> false alarm

Don't mention it. I'm grateful to have an extra pair of eyes reviewing. Better safe than sorry. :-)

Comment (pamelafox) [reply app.py:51]: lol, funny timing, I actually taught dict.get() in a Python class yesterday-
https://pamelafox.github.io/python-crash-course/week3.html#/11/0/3

**Review by pamelafox — APPROVED** [review]

File: app.py (hunk 5 above)
Comment (rayluo) [reply app.py:51]: Nice! Whom do you teach those classes to? Your course material is neat, and much more polished than [my old school attempt](https://github.com/rayluo/first-programming-book/wiki). May I borrow your materials when I and some other volunteers teach our next class?

Comment (pamelafox) [reply app.py:51]: Oh nice, an e book! I teach for corise.com, but you're free to use the content.

Lecture slides at:
https://pamelafox.github.io/python-crash-course/

Exported articles/exercises at:
https://pamelafox.github.io/python-crash-course/content/en/
(Doesn't have all the rich widgets of the corise.com platform, but the text is there.)
